modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -165,8 +164,10 @@ public void write(final ChannelHandlerContext ctx, final Object msg, final Chann
         } catch (IllegalStateException e) {
             ctx.channel().close();
         } finally {
-            if (success == false) {
-                promise.setFailure(new ClosedChannelException());
+            if (success == false && promise.isDone() == false) {
+                // The preceding failure may already have failed the promise; use tryFailure() to avoid log noise about double-completion,
+                // but also check isDone() first to avoid even constructing another exception in most cases.
+                promise.tryFailure(new ClosedChannelException());
```

Comment:
After `success` is set to `true`, the code continues to call `doWriteQueued(ctx)` which calls `doWrite` again. Is it possible for `doWrite` to throw and cause the promise of the top element polled from `outboundHoldingQueue` _not_ being completed?

Reply:
I think `doWrite()` must always complete its promise. The exception handling in this method should be only about the `ISE` that can get thrown when the pipeline is full; this code is generally designed to avoid exceptions and rely on promise completion throughout.

Reply:
Thanks David! I was going to ask whether we should move `doWrite` and the following bits out of the try-catch-finally section in that case. You beat me to it. 👍